lightning/src/util/sweep.rs

```diff
@@ -0,0 +1,887 @@
+			assert_eq!(state_lock.best_block.height, height - 1,
+				"Blocks must be connected in chain-order - the connected block height must be one greater than the previous height");
+
+			self.transactions_confirmed_internal(&mut *state_lock, header, txdata, height);
+			spending_tx_opt = self.best_block_updated_internal(&mut *state_lock, header, height);
+
+			self.persist_state(&*state_lock).unwrap_or_else(|e| {
+				log_error!(self.logger, "Error persisting OutputSweeper: {:?}", e);
+			});
+		}
+
+		if let Some(spending_tx) = spending_tx_opt {
+			self.broadcaster.broadcast_transactions(&[&spending_tx]);
+		}
+	}
+
+	fn block_disconnected(&self, header: &Header, height: u32) {
+		let mut state_lock = self.sweeper_state.lock().unwrap();
+
+		let new_height = height - 1;
+		let block_hash = header.block_hash();
+
+		assert_eq!(state_lock.best_block.block_hash, block_hash,
+		"Blocks must be disconnected in chain-order - the disconnected header must be the last connected header");
+		assert_eq!(state_lock.best_block.height, height,
+			"Blocks must be disconnected in chain-order - the disconnected block must have the correct height");
+		state_lock.best_block = BestBlock::new(header.prev_blockhash, new_height);
+
+		for output_info in state_lock.outputs.iter_mut() {
+			if output_info.status.confirmation_hash() == Some(block_hash) {
+				debug_assert_eq!(output_info.status.confirmation_height(), Some(height));
+				output_info.status.unconfirmed();
+			}
+		}
+
+		self.persist_state(&*state_lock).unwrap_or_else(|e| {
+			log_error!(self.logger, "Error persisting OutputSweeper: {:?}", e);
+		});
+	}
+}
+
+impl<B: Deref, D: Deref, E: Deref, F: Deref, K: Deref, L: Deref, O: Deref> Confirm
+	for OutputSweeper<B, D, E, F, K, L, O>
+where
+	B::Target: BroadcasterInterface,
+	D::Target: ChangeDestinationSource,
+	E::Target: FeeEstimator,
+	F::Target: Filter + Sync + Send,
+	K::Target: KVStore,
+	L::Target: Logger,
+	O::Target: OutputSpender,
+{
+	fn transactions_confirmed(
+		&self, header: &Header, txdata: &chain::transaction::TransactionData, height: u32,
+	) {
+		let mut state_lock = self.sweeper_state.lock().unwrap();
+		self.transactions_confirmed_internal(&mut *state_lock, header, txdata, height);
+		self.persist_state(&*state_lock).unwrap_or_else(|e| {
+			log_error!(self.logger, "Error persisting OutputSweeper: {:?}", e);
+		});
+	}
+
+	fn transaction_unconfirmed(&self, txid: &Txid) {
+		let mut state_lock = self.sweeper_state.lock().unwrap();
+
+		// Get what height was unconfirmed.
+		let unconf_height = state_lock
+			.outputs
+			.iter()
+			.find(|o| o.status.latest_spending_tx().map(|tx| tx.txid()) == Some(*txid))
+			.and_then(|o| o.status.confirmation_height());
+
+		// Unconfirm all >= this height.
+		let mut should_persist = false;
+		state_lock
+			.outputs
+			.iter_mut()
+			.filter(|o| o.status.confirmation_height() >= unconf_height)
+			.for_each(|o| {
+				should_persist = true;
+				o.status.unconfirmed()
+			});
```

Comment:
I don't think the logic is correct here. If `unconf_height` is `None`, then all outputs will pass the filter and be unconfirmed.

```rust
assert!(None::<u32> >= None::<u32>);
assert!(Some(1u32) >= None);
```


For example, the first call for a `Txid` at a given block height would remove all outputs confirmed at that height. Then a second call for a `Txid` at the same block height would give `None` for `unconf_height`, which would result in unconfirming all outputs *regardless of height*.

Instead, we should short-circuit if `unconf_height` is `None`. Then we wouldn't need the `should_persist` variable, either. Could we also add some test coverage for this?